lib/exename.cpp

```diff
@@ -16,58 +16,75 @@ limitations under the License.
 
 #include "exename.h"
 
-#include <limits.h>
-#include <stdlib.h>
-#include <string.h>
-#include <unistd.h>
-
-#include <sys/stat.h>
-#include <sys/types.h>
-
-#include "exceptions.h"
+#include <array>
+#include <cstring>
+#include <filesystem>
 
 namespace P4 {
 
-template <size_t N>
-static void convertToAbsPath(const char *const relPath, char (&output)[N]) {
-    output[0] = '\0';  // Default to the empty string, indicating failure.
+#if defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__) || defined(__NetBSD__)
+#include <unistd.h>
+
+#include <climits>
+#ifdef __APPLE__
+#include <mach-o/dyld.h>
+#endif
+#elif defined(_WIN32)
+#include <windows.h>
+#endif
 
-    char cwd[PATH_MAX];
-    if (!getcwd(cwd, sizeof(cwd))) return;
-    const size_t cwdLen = strlen(cwd);
-    if (cwdLen == 0) return;
-    const char *separator = cwd[cwdLen - 1] == '/' ? "" : "/";
+std::filesystem::path getExecutablePath(const std::filesystem::path &suggestedPath) {
+#if defined(__linux__) || defined(__FreeBSD__) || defined(__NetBSD__)
+    // Find the path of the executable.  We use a number of techniques that may fail or work on
+    // different systems, and take the first working one we find.  Fallback to not overriding the
+    // compiled-in installation path.
+    std::array<std::string, 4> paths = {
+        "/proc/self/exe",         // Linux
+        "/proc/curproc/file",     // FreeBSD
+        "/proc/curproc/exe",      // NetBSD
+        "/proc/self/path/a.out",  // others?
+    };
 
-    // Construct an absolute path. We're assuming that @relPath is relative to
-    // the current working directory.
-    int n = snprintf(output, N, "%s%s%s", cwd, separator, relPath);
-    BUG_CHECK(n >= 0, "Pathname too long");
+    for (const auto &path : paths) {
+        try {
+            // std::filesystem::canonical will throw on error if the path is invalid.
+            // It will also try to resolve symlinks.
+            return std::filesystem::canonical(path);
+        } catch (const std::filesystem::filesystem_error &) {
+            // Ignore and try the next path.
+        }
+    }
+#elif defined(__APPLE__)
+    static std::array<char, PATH_MAX> buffer{};
+    uint32_t size = static_cast<uint32_t>(buffer.size());
+    if (_NSGetExecutablePath(buffer.data(), &size) == 0) {
+        // TODO: What to do about the allocation here?
+        return (buffer.data());
+    }
+#elif defined(_WIN32)
+    static std::array<char, PATH_MAX> buffer{};
+    // TODO: Do we need to support this?
+    DWORD size = GetModuleFileNameA(nullptr, buffer.data(), static_cast<DWORD>(buffer.size()));
+    if (size > 0 && size < buffer.size()) {
+        return (buffer.data());
+    }
+#endif
+    // If the above fails, try to convert argv0 to a path.
+    if (std::filesystem::exists(suggestedPath)) {
+        return suggestedPath;
```

Comment:
What happens here if `suggestedPath` is a relative path?  Previously we would prepend `$CWD` to make an absolute path.